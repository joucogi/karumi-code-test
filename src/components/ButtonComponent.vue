<template>
  <div>
    <b-button :id="id" size="lg" variant="primary" @click="$emit('click')">
      <span> {{ name }} </span>
    </b-button>
  </div>
</template>
<script>
export default {
  props: {
    id: {
      type: String,
      default: "btnId",
      required: false
    },
    name: {
      type: String,
      default: "Send",
      required: false
    }
  }
};
</script>
<style scoped>
div {
  text-align: left;
}
span {
  font-weight: bold;
}
</style>
