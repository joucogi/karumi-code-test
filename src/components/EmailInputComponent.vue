<template>
  <InputComponent type="email" :placeholder="placeholder" />
</template>
<script>
import InputComponent from "./InputComponent";
export default {
  components: { InputComponent },
  props: {
    placeholder: {
      type: String,
      default: "",
      required: false
    }
  }
};
</script>
