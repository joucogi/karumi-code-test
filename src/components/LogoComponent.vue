<template>
  <div>
    <img alt="Karumi logo" src="../assets/logo.png" />
  </div>
</template>
<style scoped>
  img {
    width: 60%;
    min-width: 80px;
  }
</style>