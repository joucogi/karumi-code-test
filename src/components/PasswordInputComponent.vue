<template>
  <InputComponent
    :id="id"
    v-model="model"
    type="password"
    :placeholder="placeholder"
    @input="onInput"
  />
</template>
<script>
import InputComponent from "./InputComponent";

export default {
  components: { InputComponent },
  props: {
    id: {
      type: String,
      default: "",
      required: false
    },
    placeholder: {
      type: String,
      default: "",
      required: false
    },
    value: {
      type: String
    }
  },
  data() {
    return {
      model: this.value
    };
  },
  methods: {
    onInput(payload) {
      this.$emit("input", payload);
    }
  }
};
</script>
