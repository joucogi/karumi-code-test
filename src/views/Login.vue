<template>
  <div class="login">
    <div class="logo">
      <LogoComponent />
    </div>
    <div class="content">
      <EmailInputComponent
        v-model="username"
        id="username"
        class="mb-3"
        placeholder="Username"
        @input="setUsername"
      />

      <PasswordInputComponent
        v-model="password"
        id="password"
        class="mb-5"
        placeholder="Password"
        @input="setPassword"
      />
      <span class="mb-5 error" :class="{ hidden: !showError }">
        {{ error }}
      </span>
      <ButtonComponent id="btnLogin" name="Login" @click="loginUser()" />
    </div>
  </div>
</template>

<script>
import LogoComponent from "@/components/LogoComponent";
import EmailInputComponent from "@/components/EmailInputComponent";
import PasswordInputComponent from "@/components/PasswordInputComponent";
import ButtonComponent from "@/components/ButtonComponent";

export default {
  name: "Login",
  inject: ["login"],
  data() {
    return {
      username: "",
      password: "",
      showError: false,
      error: "The username or password youâ€™ve entered is incorrect"
    };
  },
  components: {
    LogoComponent,
    EmailInputComponent,
    PasswordInputComponent,
    ButtonComponent
  },
  methods: {
    setUsername: function(payload) {
      this.username = payload;
    },
    setPassword: function(payload) {
      this.password = payload;
    },
    loginUser: function() {
      this.showError = false;
      if (!this.login.validate(this.username, this.password)) {
        this.showError = true;
        return;
      }

      this.$router.push({ name: "Home" });
    }
  }
};
</script>
<style scoped>
.login {
  width: 100%;
  max-width: 500px;
  display: flex;
  flex-direction: column;
}

.login > div {
  flex: 1;
}

.logo {
  display: flex;
  align-items: center;
  justify-content: center;
  align-content: center;
}

.content {
  display: flex;
  flex-direction: column;
}

.hidden {
  display: none;
}

.error {
  color: red;
  font-size: 12px;
}
</style>
