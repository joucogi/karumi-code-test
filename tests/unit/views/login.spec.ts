import { shallowMount, Wrapper } from "@vue/test-utils";
import Vue from "vue";

import Login from "@/views/Login.vue";
import LogoComponent from "@/components/LogoComponent.vue";
import EmailInputComponent from "@/components/EmailInputComponent.vue";
import PasswordInputComponent from "@/components/PasswordInputComponent.vue";
import ButtonComponent from "@/components/ButtonComponent.vue";
import { toMatchSnapshot, componentExists } from "../helpers";

import LoginUserMock from "../mocks/login-user-mock";

let wrapper: Wrapper<Vue>;
const login = new LoginUserMock();

beforeEach(() => {
  wrapper = shallowMount(Login, {
    stubs: ["b-form-input", "b-button"],
    provide: {
      login
    }
  });
});

describe("Login View", () => {
  it("contains a LogoComponent", () => {
    componentExists(wrapper, LogoComponent);
  });

  it("contains an EmailInputComponent", () => {
    componentExists(wrapper, EmailInputComponent);
  });

  it("contains a PasswordEmailInputComponent", () => {
    componentExists(wrapper, PasswordInputComponent);
  });

  it("contains a ButtonComponent", () => {
    componentExists(wrapper, ButtonComponent);
  });

  it("contains an span tag with error message", () => {
    const span = wrapper.find("span");
    expect(span.text()).toContain(
      "The username or password youâ€™ve entered is incorrect"
    );
  });

  it("when button is clicked validate method from login service is called", async () => {
    const button = wrapper.find("#btnLogin");
    await button.vm.$emit("click");
    expect(login.validate.mock.calls.length).toBe(1);
  });

  it("renders correctly", () => {
    toMatchSnapshot(wrapper);
  });
});
