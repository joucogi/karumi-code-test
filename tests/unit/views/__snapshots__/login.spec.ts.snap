// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login View renders correctly 1`] = `
<div
  class="login"
>
  <div
    class="logo"
  >
    <logocomponent-stub />
  </div>
   
  <div
    class="content"
  >
    <emailinputcomponent-stub
      class="mb-3"
      id="username"
      placeholder="Username"
      value=""
    />
     
    <passwordinputcomponent-stub
      class="mb-5"
      id="password"
      placeholder="Password"
      value=""
    />
     
    <span
      class="mb-5 error hidden"
    >
      
      The username or password youâ€™ve entered is incorrect
    
    </span>
     
    <buttoncomponent-stub
      id="btnLogin"
      name="Login"
    />
  </div>
</div>
`;
